### PR TITLE
allow type params before navigation suffix

### PR DESCRIPTION
**Pull request by brandonspark**
In my pursuits running the Kotlin parser on many open-source repos, I've noticed that a very common we cannot parse is expressions of the form `Foo<*>::bar`. It seems that this is an example of suffixing an expression with type parameters, and then doing a navigation suffix.

Unfortunately, it seems that naively writing the grammar with type parameters after arbitrary expressions is a surefire way to break comparison expressions. I decided to follow the same form as `call_suffix`, and instead allow type parameters in only a particular suffix, that being navigation suffixes.

Now, it works. Our parse rate went up like a percent as a result.

Test plan: Automated tests.

### CODE REVIEW

**Comment by RenFraser** [comment]
This is awesome @brandonspark. I sincerely hope that you keep contributing! I thought that this repo was losing traction but it seems I was wrong! Thank you @fwcd and @brandonspark! :)

**Comment by fwcd** [comment]
Thanks for investigating! Please rebase it onto main (as always), so we can run CI.

File: grammar.js [hunk 1]
```diff
@@ -673,6 +674,8 @@ module.exports = grammar({
     indexing_suffix: $ => seq("[", sep1($._expression, ","), "]"),
 
     navigation_suffix: $ => seq(
+      // this introduces ambiguities with 'less than' for comparisons
+      optional($.type_arguments),
```
Comment (fwcd) [inline grammar.js:678]: Hm, I'm not quite sure how I feel about this workaround. While I understand that it's hard to parse, it seems to deviate quite a bit from the [reference grammar](https://kotlinlang.org/docs/reference/grammar.html#navigationSuffix) and including the type arguments in the navigation suffix node (both of which are exposed), downstream consumers might have to rewrite the syntax tree quite a bit to get accurate results.

Comment (brandonspark) [reply grammar.js:678]: I suppose it just comes down to how much you weigh being able to parse this versus the compromises we've already made with the syntax tree. For instance, this `call_suffix` workaround (which basically does the same thing, by embedding the optional type arguments in a different node of the tree) was worthy enough of being included, so I see this as adding just another case.

I can keep thinking about it, but this seems like a rather pervasive parsing issue, from the things I've seen, and I'm not sure if we can make something truly faithful to the original syntax specification. At the end of the day, the Kotlin grammar is a reference for specification, but implementation may face some difficulties emulating it.
